assets/js/plan_template.js

```diff
@@ -0,0 +1,175 @@
+// JS for plan_template.html
+import { logOut, updateUsername } from "./user.js";
+document.getElementById("logout-confirm-btn").addEventListener(
+    "click", logOut
+)
+
+const username = updateUsername();
+document.getElementById("profile").addEventListener("click", () => window.location = `/v1/users/${username}/profile`);
+
+function removeLastRow() {
+    const table = document.getElementById("template");
+    // prevent deleting table headers
+    if (table.rows.length > 1) {
+        table.deleteRow(-1);
+    }
+}
+
+function insertNewRow(start = '8', end = '10', category = 'Visit') {
+    const template = document.getElementById("template");
+    let newRow = template.insertRow(-1);
+    let newCell = newRow.insertCell(0);
+    let newCell2 = newRow.insertCell(1);
+    let newCell3 = newRow.insertCell(2);
+
+    let select = document.createElement("select");
+    let cafeOption = document.createElement("option");
+    let attractionOption = document.createElement("option");
+
+    let startTime = hourDropdown();
+    startTime.classList.add("form-select");
+    startTime.value = start;
+
+    let endTime = hourDropdown();
+    endTime.classList.add("form-select");
+    endTime.value = end;
+
+    cafeOption.text = "Eatery";
+    attractionOption.text = "Visit";
+
+    select.classList.add("form-select");
+    select.add(cafeOption);
+    select.add(attractionOption);
+    select.value = category;
+
+    newCell.append(select);
+
+    newCell2.append(startTime);
+
+    newCell3.append(endTime);
+}
+
+function hourDropdown() {
+    let select = document.createElement("select");
+    for (let hour = 8; hour < 20; hour++) {
+        let option = document.createElement("option");
+        option.text = hour.toString();
+        select.add(option);
+    }
+    return select;
+}
+
+function tableToJSON() {
+    return $("#template").tableToJSON(
+        {
+            extractor: function ($cellIdx, $cell) {
+                return $cell.find("select").val()
+            }
+        }
+    )
+}
+
+function tableToSlots() {
+    const rows = tableToJSON()
+    return rows.map(
+        function (row) {
+            return {
```

Comment:
Maybe we could consider defining a new struct here to simplify codes between line 76 and 85. 
For example, we could define a struct or class in JS `Slot = {"attr": val, ...}`, a slot constructor `makeSlot(row)`.

Reply:
Then the return of tableToSlots() can be simplified into 
`return rows.map( row => makeSlot(row) )`

Reply:
Done.